More dispatchables

bedeho opened the pull request:
- fix for `create_category`
- [x] update_category
- [x] create_thread
- [x] moderate_thread

bedeho committed:
fix: missing event from create_category

bedeho committed:
new: update_category

no tests yet!

siman commented on src/lib.rs:
```diff
@@ -363,253 +527,465 @@ struct Category<BlockNumber, Moment, AccountId> {
+
+            Self::build_category_tree_path(category_id, &mut category_tree_path);
+
+            if !category_tree_path.is_empty() {
```
Isn't it enough just check that `CategoryById` has a category by specified `category_id`?

siman replied:
you can check that category exists, and if not, the return from this function.
It will allow you to move the current block of `if` statement out of `if`.

Pseudocode:

```rust
if category_does_not_exist {
    // If path is empty, it just means the category did not exist,
    // which means the user provided an invalid category id.
    Err(ERROR_CATEGORY_DOES_NOT_EXIST)
}

// update the found category here:
// ....
```

bedeho replied:
Discarded, premature review.

siman commented on src/lib.rs:
```diff
@@ -363,253 +527,465 @@ struct Category<BlockNumber, Moment, AccountId> {
+                let mut category = category_tree_path.get_mut(0).unwrap();
+
+                // Value change events params
+                let mut archived_change_to = None;
```
You can change a signature of this function a little bit:

```rust
fn update_category(origin, category_id: CategoryId, archived_opt: Option<bool>, deleted_opt: Option<bool>) -> dispatch::Result {
    // checks...

    // start updating a category:
    if let Some(archived) = archived_opt {
        category.archived = archived;
    }

    if let Some(deleted) = deleted_opt {
        category.deleted = deleted;
    }

    if archived_opt.is_some() || deleted_opt.is_some() {
        // Write back mutated category
        <CategoryById<T>>::insert(category_id, category);

        // Generate event
        Self::deposit_event(RawEvent::CategoryUpdated(category_id, archived_opt, deleted_opt));
    }
}
```

siman commented on src/lib.rs:
```diff
@@ -363,253 +527,465 @@ struct Category<BlockNumber, Moment, AccountId> {
+                category_id: category_id,
+                thread_nr: 777777,
+                moderation : None,
+                num_unmoderated_posts: 0,
```
Did you consider a little bit different fields in Thread struct for a number of moderated / unmoderated posts?
Having `num_total_posts` is easier to digest (for human) because having `num_unmoderated_posts`  and `num_moderated_posts` will require a programmer or API consumer to sum these two numbers.

Plus `num_total_posts` does not require to update it with each moderation action, in contrast to `num_unmoderated_posts` - where you need to decrease this number with every moderation on this tread/post.

```rust
struct Thread {
     // ... other fields
    num_total_posts: u64, // <-- instead of num_unmoderated_posts
    num_moderated_posts: u64,
    // ... other fields
}

bedeho replied:
Discarded, premature review.

siman commented on src/lib.rs:
```diff
@@ -363,253 +527,465 @@ struct Category<BlockNumber, Moment, AccountId> {
+                category_id: category_id,
+                thread_nr: 777777,
+                moderation : None,
+                num_unmoderated_posts: 0,
```
Seems like a typo:
in this case, `num_unmoderated_posts` should be `1`
and `num_moderated_posts` should be `0`.

bedeho replied:
Discarded, premature review.

bedeho committed:
Big commit

Just rolling everything int one commit, as there was no partial review, and most intermediate commits were WIP without context.

Bad, I know.

siman commented on src/lib.rs:
```diff
@@ -392,16 +438,17 @@ pub struct Post<BlockNumber, Moment, AccountId> {
     
     /// Edits of post ordered chronologically by edit time.
-    post_text_edits: Vec<PostTextEdit<BlockNumber, Moment>>,
+    expired_post_texts: Vec<ExpiredPostText<BlockNumber, Moment>>,
```
Is `expired` the best word here? Maybe something like `old_post_texts` or something, just thinking. Because a text is actually **replaced** by the post author.

bedeho replied:
Assuming you are talking about both variable `expired_post_texts` and type `ExpiredPostText`. I think expired is good, because it states very clearly that its no longer valid as the post text, "old" is a bit more ambiguous, it just says it was text a long time ago.

I will change if you give me a specific suggestion you think is better, but this seems very minor.

siman replied:
What about this?
struct `PostTextChange` and `text_change_history` as a field name.

```rust
struct Post {
    // ....
    text_change_history: Vec<PostTextChange<BlockNumber, Moment>>,
    // ....
}
```

bedeho replied:
Done.

siman commented on src/lib.rs:
```diff
@@ -432,16 +479,24 @@ pub struct Thread<BlockNumber, Moment, AccountId> {
     /// `created_at` is not sufficient.
-    thread_nr: u32,
+    /// Starts at 1 for first thread in category.
+    thread_nr_in_category: u32,
```
Regarding `_nr_`
I have checked the Internet for "what is a shortcut for a number", and the answer is `no`.
https://forum.wordreference.com/threads/abbreviation-of-number-n-n%C2%B0-nr-nbr-no.264328/
https://english.stackexchange.com/questions/43095/what-is-the-correct-abbreviation-for-the-word-numbers

but I guess that `no` could be considered as "No" word (negative).

siman replied:
What about `sn_in_category` or something like this. Which means `a serial number of the thread in its category.`

siman replied:
At least I believe that `thread_` prefix could be removed from `thread_nr_in_category` because this field is already in a struct called `Thread`.

bedeho replied:
I removed `thread_` prefix and also `post_nr`.

siman commented on src/lib.rs:
```diff
@@ -476,28 +543,38 @@ pub struct Category<BlockNumber, Moment, AccountId> {
-    /// Parent category, if present, otherwise this category is a root categoryl
-    parent_id: Option<CategoryId>,
+    /// Position as child in parent, if present, otherwise this category is a root category
+    slot_in_parent_category: Option<ChildPositionInParentCategory>,
```
The structure has a word `Position` but a field has a word `slot`. Looks inconsistent.

bedeho replied:
Yes, changed variable name.

siman commented on src/lib.rs:
```diff
@@ -476,28 +543,38 @@ pub struct Category<BlockNumber, Moment, AccountId> {
 
+impl<BlockNumber, Moment, AccountId> Category<BlockNumber, Moment, AccountId>  {
+
+    fn num_threads_ever_created(&self) -> u32 {
```
I would prefer to have `num_direct_total_threads` on a category instead.

bedeho replied:
Changed to `num_threads_created` , as we agreed in the discussion.

siman commented on src/lib.rs:
```diff
@@ -663,7 +763,10 @@ decl_module! {
                 archived: false,
-                parent_id: parent,
+                num_direct_subcategories: 0,
+                num_direct_unmoderated_threads: 0,
```
Seems like a typo here:
```suggestion
                num_direct_unmoderated_threads: 1,
                num_direct_moderated_threads: 0,
```

bedeho replied:
Fixed.

siman commented on src/lib.rs:
```diff
@@ -673,18 +776,215 @@ decl_module! {
+            let new_post = Post {
+                id: new_post_id,
+                thread_id: new_thread_id,
+                post_nr_in_thread: 1, // Starts at 1
```
Would you like to create constant for a default value of any entity in a list and use it for subcategories, threads, posts?
```suggestion
// somewhere above:
const DEFAULT_POSITION_IN_LIST: u32 = 1;
// ... other code...

                post_nr_in_thread: DEFAULT_POSITION_IN_LIST, // Starts at 1
```

bedeho replied:
This issue is resolved in https://github.com/Joystream/substrate-forum-module/pull/13 , so there is no longer a constant.

siman commented on src/lib.rs:
```diff
@@ -673,18 +776,215 @@ decl_module! {
+            let path = Self::build_category_tree_path(thread.category_id);
+
+            // Path must be non-empty, as category id is from thread in state
+            assert!(path.len() > 0);
```
```suggestion
            assert!(path.is_empty());
```

bedeho replied:
Changed to `!path.is_empty()`, as its asserting non-emptyness.

bedeho committed:
Fixed minor issues based on review

mnaamani commented:
Test are passing
